src/core/border_router/routing_manager.cpp

```diff
@@ -3110,6 +3183,245 @@ void RoutingManager::RsSender::HandleTimer(void)
+        bool currentPrefixUpdated = false;
+
+        if (!IsValidOmrPrefix(mPrefix))
+        {
+            currentPrefixPreferredRemaining = 0;
+            currentPrefixValidRemaining     = 0;
+        }
+
+        VerifyOrExit(message.IsValid(), error = kErrorParse);
+
+        for (const Ip6::Nd::Option &option : message)
+        {
+            if (option.GetType() == Ip6::Nd::Option::Type::kTypePrefixInfo)
+            {
+                const Ip6::Nd::PrefixInfoOption &pio = static_cast<const Ip6::Nd::PrefixInfoOption &>(option);
+                Ip6::Prefix                      pioPrefix;
+
+                pio.GetPrefix(pioPrefix);
+
+                // Note that we SHOULD accept ULA prefix since it could be used by the internet infrastructure like
+                // NAT64.
+                if (pioPrefix.mLength > kOmrPrefixLength || pioPrefix.mLength == 0 || pioPrefix.IsLinkLocal() ||
+                    pioPrefix.IsMulticast())
+                {
+                    LogWarn("Ignore invalid platform generated prefix %s", pioPrefix.ToString().AsCString());
+                    continue;
+                }
+
+                // In some cases, the platform may delegate us more than one prefixes, usually one GUA prefix with one
+                // or more ULA prefixes, in this case, we only preserve the smallest one. Since currently GUA prefixes
+                // are allocated from the 2000::/3 block while the ULA prefixes are generated from the fc00::/7 block.
+                // Replace the current selected platform OMR prefix if the one in RA is favored over the current one.
+                if (favoredPrefixFromRa.GetLength() == 0 ||
+                    ((raPrefixPreferredRemaining == 0 || pioPrefix < favoredPrefixFromRa) &&
+                     pio.GetPreferredLifetime() > 0))
+                {
+                    raHasValidPio = true;
+
+                    // From RFC4861: MaxRtrAdvInterval: The maximum time allowed between sending unsolicited multicast
+                    // Router Advertisements from the interface, in seconds.  MUST be no less than 4 seconds and no
+                    // greater than 1800 seconds.
+                    //
+                    // Default: 600 seconds
+                    //
+                    // So it is safe to assume that the platform will generate another RA message for the Thread
+                    // interface within 30 minutes (so the timeout will be refreshed), thus it is safe to set the
+                    // timeout to UINT16_MAX to avoid overflow.
+
+                    favoredPrefixFromRa = pioPrefix;
+                    raPrefixPreferredRemaining =
+                        Min(pio.GetPreferredLifetime(), static_cast<uint32_t>(NumericLimits<uint16_t>::kMax));
+                    raPrefixValidRemaining =
+                        Min(pio.GetValidLifetime(), static_cast<uint32_t>(NumericLimits<uint16_t>::kMax));
+                }
```

Comment:
One idea/suggestion here:

We can consider storing the DHCP6 PD related PIO in `PdPrefixManager` as a `DiscoveredPrefixTable::Entry`. 
- This way we can use all existing helper methods of `Entry`, like `SetFrom(const Ip6::Nd::PrefixInfoOption &aPio)` or `GetExpireTime(void)` or `GetStaleTime()` or `IsDeprecated()`?
- We may need to declare `PdPrefixManager` as a `friend` of `DiscoveredPrefixTable` and its `Entry`.